### PR TITLE
Add the simple diffusion model tutorial

### PR DESCRIPTION
**Pull request by lucianommartins**
Including a simple diffusion model development and training tutorial.

### CODE REVIEW

File: docs/digits_diffusion_model.md [hunk 1]
```diff
@@ -0,0 +1,822 @@
...
+    if epoch % 100 == 0:
+        print(f"Epoch {epoch}, Loss: {moving_avg_loss:.4f}")
+
+print("\nTraining completed.")
+```
+
++++ {"id": "SHX69WgxKUs7"}
+
+
+
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:623]:
```suggestion
```

File: docs/digits_diffusion_model.md [hunk 2]
```diff
@@ -0,0 +1,822 @@
...
++++ {"id": "s-iqch4HKlBV"}
+
+## Training loss visualization
+
+It uses a logarithmic scale to better display the exponential decay of the loss values over time. This representation helps identify both early rapid improvements and later fine-tuning phases of the training process.
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:628]:
```suggestion
We use a logarithmic scale to better display the exponential decay of the loss values over time. This representation helps identify both early rapid improvements and later fine-tuning phases of the training process.
```

File: docs/digits_diffusion_model.md [hunk 3]
```diff
@@ -0,0 +1,822 @@
...
+plt.figure(figsize=(10, 5))            # Create figure with wide aspect ratio for clarity
+plt.plot(losses)                       # losses: List[float] - historical EMA loss values
+plt.title('Training Loss Over Time')
+plt.xlabel('Epoch')
+plt.ylabel('Loss')
+plt.yscale('log')                      # Use log scale to better visualize exponential decay
+plt.grid(True)                         # Add grid for easier value reading
+plt.show()
+```
+
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:648]: Suggestion: add some comment explaining what we can see from this visualization (i.e., "We can see that our model performs well in this dataset, due to the flat curve in the right-most part of the graph.")

File: docs/digits_diffusion_model.md [hunk 4]
```diff
@@ -0,0 +1,822 @@
...
++++ {"id": "M2ql0KwYJLqn"}
+
+## Visualization Functions
+
+Includes utilities for:
+- Sample generation
+- Forward/reverse process visualization
+- Training progress tracking
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:656]:
```suggestion
In this section, we include utilities for:
- Sample generation
- Forward/reverse process visualization

These tools can help you track the training progress for your model.
```

I didn't see a "Training progress tracking" code cell but maybe you mean something different, such as saying that the sample generation and forward/reverse process visualization help you track the training progress. In that case I would spell it out and not add it as a bullet item.

File: docs/digits_diffusion_model.md [hunk 5]
```diff
@@ -0,0 +1,822 @@
...
+# Generate visualization
+key, subkey = jax.random.split(key)
+print("\nFull Forward and Reverse Process:")
+plot_forward_and_reverse(model, diffusion, images_test[0], subkey)
+```
+
++++ {"id": "o43bRWpiM6Mt"}
+
+## Summary
+This tutorial demonstrated the implementation of a diffusion model using JAX and Flax libraries. You explored the U-Net architecture with attention mechanisms, efficient training strategies using JIT compilation, and comprehensive visualization techniques for the diffusion process. The implementation showcases JAX's powerful features for high-performance machine learning, including automatic differentiation, vectorization, and just-in-time compilation.
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:820]:
```suggestion
This tutorial demonstrated the implementation of a diffusion model using JAX and Flax libraries. We explored the U-Net architecture with attention mechanisms, efficient training strategies using JIT compilation, and comprehensive visualization techniques for the diffusion process. The implementation showcases JAX's powerful features for high-performance machine learning, including automatic differentiation, vectorization, and just-in-time compilation.
```

File: docs/digits_diffusion_model.md [hunk 6]
```diff
@@ -0,0 +1,822 @@
...
+print("\nTraining completed.")
+```
+
++++ {"id": "SHX69WgxKUs7"}
+
+
+
++++ {"id": "s-iqch4HKlBV"}
+
+## Training loss visualization
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:626]:
```suggestion
### Training loss visualization
```

File: docs/digits_diffusion_model.md [hunk 7]
```diff
@@ -0,0 +1,822 @@
...
+
+# Image generation in JAX: a simple diffusion model
+
+[![Open in Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/jax-ml/jax-ai-stack/blob/main/docs/digits_diffusion_model.ipynb)
+
+In [Debugging in JAX: a Variational autoencoder (VAE) model](https://colab.research.google.com/github/jax-ml/jax-ai-stack/blob/main/docs/digits_vae.ipynb) you explored a simplified version of a [Variational Autoencoder (VAE)](https://en.wikipedia.org/wiki/Variational_autoencoder) trained on the simple digits data. In this tutorial you will find the steps to develop, train and perform inferences with a simple diffusion model developed with JAX, Flax, NNX and Optax. It includes:
+- preparing the dataset
+- Developing the custom diffusion model
+- Creating the loss and training functions
+- Perform the model training using Colab TPU v2 as a hardware accelerator
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:23]:
```suggestion
- Perform the model training using Colab TPU v2 as a hardware accelerator
- Visualizing and tracking your model progress.
```

File: docs/digits_diffusion_model.md [hunk 8]
```diff
@@ -0,0 +1,822 @@
...
+plt.show()
+```
+
++++ {"id": "exKxj9OcG0yk"}
+
+## Building your diffusion model
+
+Now you will start developing the parts of your [diffusion model](https://en.wikipedia.org/wiki/Diffusion_model). It if composed of the backbone part in a (UNet architecture)[https://en.wikipedia.org/wiki/U-Net] and the diffusion layers.
+
+## U-Net Architecture
```
Comment (melissawm) [inline docs/digits_diffusion_model.md:119]:
```suggestion
### U-Net Architecture
```

**Review by melissawm — COMMENTED** [review]
Hi all - left a few comments, some typos and a few suggestions.

I also included a suggestion for a sections/subsections hierarchy that I think makes the flow of the tutorial better. Cheers!

**Comment by melissawm** [comment]
I believe adding the .md file to https://github.com/jax-ml/jax-ai-stack/blob/977d5c4759160c3aa0e784a9d6afb07e32900f9b/docs/conf.py#L32 will fix the Read the Docs build.

**Review by jakevdp — APPROVED** [review]
Looks good overall! A few little book-keeping things:

- You'll need to ensure the ipynb and markdown are linked via jupytext. See https://jax-ai-stack.readthedocs.io/en/latest/contributing.html#adding-a-new-notebook for details.
- To make the new tutorial render as part of the documentation, you need to add it to the table of contents in https://github.com/jax-ml/jax-ai-stack/blob/main/docs/tutorials.md
- You'll need to add the markdown file here to avoid a build failure: https://github.com/jax-ml/jax-ai-stack/blob/0392327d3d689b56d48dc4d90aa50e582b72f4bf/docs/conf.py#L40
- You'll probably need to add the notebook to the exclude patterns unless you want the doc build to attempt notebook execution: https://github.com/jax-ml/jax-ai-stack/blob/0392327d3d689b56d48dc4d90aa50e582b72f4bf/docs/conf.py#L63

Once those are addressed, we'll be able to see this as part of a build, which should make it easier to review. Thanks!

File: docs/digits_diffusion_model.md (hunk 2 above)
Comment (lucianommartins) [reply docs/digits_diffusion_model.md:625]: As in the common Google tech writing guideline, I do prefer to keep it without the "We language".

File: docs/digits_diffusion_model.md (hunk 4 above)
Comment (lucianommartins) [reply docs/digits_diffusion_model.md:653]: I'm adjusting part of it - As in the common Google tech writing guideline, I do prefer to keep it without the "We language".

File: docs/digits_diffusion_model.md (hunk 5 above)
Comment (lucianommartins) [reply docs/digits_diffusion_model.md:817]: As in the common Google tech writing guideline, I do prefer to keep it without the "We language".

File: docs/digits_diffusion_model.md [hunk 9]
```diff
@@ -0,0 +1,822 @@
...
+        # Output layers
+        self.final_norm = nnx.LayerNorm(features, rngs=rngs)
+        self.final_conv = nnx.Conv(in_features=features,
+                                 out_features=out_channels,
+                                 kernel_size=(3, 3),
+                                 strides=(1, 1),
+                                 padding=((1, 1), (1, 1)),
+                                 rngs=rngs)
+
+    def _create_attention_block(self, channels: int, rngs: nnx.Rngs) -> Callable:
```
[the remark this answers is not in the document]
Comment (lucianommartins) [reply docs/digits_diffusion_model.md:184]: Hi @vfdev-5 - It is possible, but I preferred to create everything as close to "from scratch" as possible.

File: docs/digits_diffusion_model.md [hunk 10]
```diff
@@ -0,0 +1,821 @@
+---
+jupytext:
```
Comment (jakevdp) [inline docs/digits_diffusion_model.md:2]: There should be a `formats` entry here. Did you sync the ipynb and md sources? https://jax-ai-stack.readthedocs.io/en/latest/contributing.html#adding-a-new-notebook

Comment (jakevdp) [reply docs/digits_diffusion_model.md:2]: This is important becuase otherwise the `md` and `ipynb` may have mismatched content.
